qiskit/pulse/logical_elements_frames.py

```diff
@@ -0,0 +1,499 @@
+        """
+        for qubit_index in index:
+            if not isinstance(qubit_index, (int, np.integer)) or qubit_index < 0:
+                raise PulseError("Both indices of coupled qubits must be non-negative integers")
+
+    @property
+    def name(self) -> str:
+        """Return the name of this coupler"""
+        return f"Coupler{self.index}"
+
+
+class Frame:
+    """Base class for pulse module frame.
+
+    Because pulses used in Quantum HW are typically AC pulses, the carrier frequency and phase
+    must be defined. The ``Frame`` is the object which sets the frequency and phase for the carrier,
+    and each pulse is associated with a frame.
+
+    The different types of frames dictate how the frequency and phase duo are defined:
+
+    - ``GenericFrame`` is used to custom frames, where the frequency is defined by the user.
+    - ``QubitFrame`` is associated with the default driving frequency of a qubit.
+    - ``MeasurementFrame`` is associated with the default measurement frequency of a qubit.
+
+    Instructions on ``Frame``s like set/shift frequency/phase are broadcasted to every ``MixedFrame``
+    which involves the same ``Frame``. The default initial phase for every frame is 0.
+    """
+
+    def __init__(self, name: str):
+        """Create ``Frame``.
+
+        Args:
+            name: A unique identifier used to identify the frame.
+        """
+        self._name = name
+        self._hash = hash(name)
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """Return the name of the frame."""
+        pass
+
+    def __repr__(self) -> str:
+        return self.name
+
+    def __eq__(self, other: "Frame") -> bool:
+        """Return True iff self and other are equal, specifically, iff they have the same type and name.
+
+        Args:
+            other: The frame to compare to this one.
+
+        Returns:
+            True iff equal.
+        """
+        return type(self) is type(other) and self._name == other._name
+
+    def __hash__(self) -> int:
+        return self._hash
+
+
+class GenericFrame(Frame):
+    """Pulse module GenericFrame.
+
+    The ``GenericFrame`` is used for custom user defined frames, which are not associated with any
+    backend defaults. It is especially useful when the frame doesn't correspond to any frame of
+    the typical qubit model, like qudit control for example.
+
+    ``GenericFrame``s are identified by their unique name.
+
+    """
+
+    def __init__(self, name: str, frequency: float, phase: Optional[float] = 0.0):
+        """Create ``Frame``.
+
+        Args:
+            name: A unique identifier used to identify the frame.
+            frequency: The initial frequency set for the frame.
+            phase: The initial phase set for the frame. Default value 0.
+
+        """
+        self._frequency = frequency
+        self._phase = phase
+        super().__init__(name)
+
+    @property
+    def name(self) -> str:
```

Comment:
In the sense that this object can specify an NCO, it might be better to call this `identifier`? For example, some backend may provide custom Frame, and administrator is not interested in contributing to Qiskit. Then one can use `GenericFrame` as an alias of a particular hardware NCO. To me `name` sounds like something end-users can freely define, but indeed mapping to NCO is necessary to run it on hardware.